plugins/hls-hlint-plugin/src/Ide/Plugin/Hlint.hs

```diff
@@ -0,0 +1,375 @@
+  (
+    descriptor
+  --, provider
+  ) where
+import Refact.Apply
+import Control.Arrow ((&&&))
+import Control.DeepSeq
+import Control.Exception
+import Control.Lens ((^.))
+import Control.Monad
+import Control.Monad.IO.Class
+import Control.Monad.Trans.Except
+import Data.Aeson.Types (ToJSON(..), FromJSON(..), Value(..))
+import Data.Binary
+import Data.Hashable
+import qualified Data.HashMap.Strict as Map
+import Data.Maybe
+import qualified Data.Text as T
+import qualified Data.Text.IO as T
+import Data.Typeable
+import Development.IDE.Core.FileStore
+import Development.IDE.Core.OfInterest
+import Development.IDE.Core.Rules
+import Development.IDE.Core.Shake hiding (withIndefiniteProgress)
+import Development.IDE.Types.Diagnostics as D
+import Development.IDE.Types.Location
+import Development.Shake
+
+#ifdef GHC_LIB
+import Development.IDE.Core.RuleTypes (GhcSession(..))
+import Development.IDE.GHC.Util (hscEnv)
+import "ghc-lib" GHC hiding (DynFlags(..))
+import "ghc" GHC as RealGHC (DynFlags(..))
+import "ghc" HscTypes as RealGHC.HscTypes (hsc_dflags)
+import qualified "ghc" EnumSet as EnumSet
+import Language.Haskell.GhclibParserEx.GHC.Driver.Session as GhclibParserEx (readExtension)
+#else
+import GHC hiding (DynFlags(..))
+#endif
+
+import Ide.Logger
+import Ide.Types
+import Ide.Plugin
+import Ide.Plugin.Config
+import Ide.PluginUtils
+import Language.Haskell.HLint as Hlint
+import Language.Haskell.LSP.Core
+import Language.Haskell.LSP.Types
+import qualified Language.Haskell.LSP.Types      as LSP
+import qualified Language.Haskell.LSP.Types.Lens as LSP
+import System.FilePath (takeFileName)
+import System.IO (hPutStr, noNewlineTranslation, hSetNewlineMode, utf8, hSetEncoding, IOMode(WriteMode), withFile, hClose)
+import System.IO.Temp
+import Text.Regex.TDFA.Text()
+import GHC.Generics (Generic)
+
+-- ---------------------------------------------------------------------
+
+descriptor :: PluginId -> PluginDescriptor
+descriptor plId = (defaultPluginDescriptor plId)
+  { pluginRules = rules
+  , pluginCommands =
+      [ PluginCommand "applyOne" "Apply a single hint" applyOneCmd
+      , PluginCommand "applyAll" "Apply all hints to the file" applyAllCmd
+      ]
+    , pluginCodeActionProvider = Just codeActionProvider
+  }
+
+data GetHlintDiagnostics = GetHlintDiagnostics
+    deriving (Eq, Show, Typeable, Generic)
+instance Hashable GetHlintDiagnostics
+instance NFData   GetHlintDiagnostics
+instance Binary   GetHlintDiagnostics
+
+type instance RuleResult GetHlintDiagnostics = ()
+
+rules :: Rules ()
+rules = do
+  define $ \GetHlintDiagnostics file -> do
+    hlintOn' <- hlintOn <$> getClientConfigAction
+    logm $ "hlint:rules:hlintOn=" <> show hlintOn'
+    ideas <- if hlintOn' then getIdeas file else return (Right [])
+    return (diagnostics file ideas, Just ())
+
+  getHlintSettingsRule (HlintEnabled [])
+
+  action $ do
+    files <- getFilesOfInterest
+    void $ uses GetHlintDiagnostics $ Map.keys files
```

Comment:
Add a comment explaining what this action is for

Reply:
Does this even work? Are hlints refereshed when a file is edited?

Reply:
Well i can respond yes for the first two questions quickly, but you have to believe me or try the branch 😉 
But i catch your point, maybe a test case would be great.
The comment should describe how `GetHlintDiagnostics` uses `getIdeas` that in turn depends on either `getParsedModule` (if the hls ghc version matches the used one by hlint) or `getFileContents`. They are the source of hlint changes and the connection with the shake graph.

Reply:
Added a comment explaining how hlint diagnostics could be recomputed

Reply:
> But i catch your point, maybe a test case would be great.

Provided evidence that file edits refresh hlint diagnostics with 61df7a6 and explained how it is done in comments with 
cb0d25d